internal/server/polling/server.go

```diff
@@ -97,9 +99,86 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 		return fmt.Errorf("failed to list pull requests: %w", err)
 	}
 
+	// Log the PRs.
 	for _, pr := range prs {
 		s.log.Info("pull request", "pr", pr)
 	}
 
+	annotations := map[string]string{
+		bbp.AnnotationBBPKey: bbp.AnnotationBBPValue,
+	}
+
+	// List Terraform objects, created by BBP.
+	tfList, err := s.listTerraformObjects(ctx, tf, annotations)
+	if err != nil {
+		return fmt.Errorf("failed to list Terraform objects: %w", err)
+	}
+
+	// Create a list of objects to delete.
+	oldTFs := []v1alpha2.Terraform{}
+
+	for _, branchTF := range tfList.Items {
+		branchSource, err := s.getSource(ctx, &branchTF)
+		if err != nil {
+			s.log.Error(err, "failed to get Source")
+			continue
+		}
+
+		shouldDeleteTF := true
+
+		for _, pr := range prs {
+			if pr.HeadBranch == branchSource.Spec.Reference.Branch && fmt.Sprintf("%d", pr.Number) == branchSource.Annotations[bbp.AnnotationPRIDKey] {
+				shouldDeleteTF = false
+				break
+			}
+		}
+
+		if shouldDeleteTF {
+			oldTFs = append(oldTFs, branchTF)
+		}
+	}
+
+	// Delete TF objects without corresponding PRs.
+	for _, oldTF := range oldTFs {
+		err = s.deleteTerraform(ctx, &oldTF)
+
+		if err != nil {
+			s.log.Error(err, "failed to delete Terraform object")
+		}
+	}
+
+	// Create a list of objects to create.
+	newPRs := []provider.PullRequest{}
+
+	// Process the PRs.
+	for _, pr := range prs {
+		shouldAddPR := true
+
+		for _, branchTF := range tfList.Items {
+			branchSource, err := s.getSource(ctx, &branchTF)
+			if err != nil {
+				continue
+			}
+
+			if pr.HeadBranch == branchSource.Spec.Reference.Branch {
```

Comment:
Isn't there an annotation for the PR being tracked? Are there any ways the test in this line could succeed when you don't want it to?

Reply:
@squaremo do you mean to use PR ID instead of branch name here? Can do it, thanks.